### ISSUE TITLE
Visual Studio crashes if SonarLint is enabled (memory consumption)

### ISSUE THREAD

**Issue by bamatroot** [SonarSource/sonarlint-visualstudio#2214]
### Description
Visual Studio crashes if the SonarLint extension is enabled. Memory consumption of VS increases heavily during build. 

| Action             | SonarLint disabled                                           | SonarLint enabled |
| ---------------- | ------------------------------------------------------------ | ----------------- |
| Open Solution          | ~400 MB | ~1500MB            |
| Rebuild Solution | ~1500MB                                       | ~4100MB    |

- there are no issues in the SonarLint Output
`Initializing the daemon package...
[Open in IDE] Creating request listener...
[Open in IDE] Checking availability of port 11111
[Open in IDE] Request listener created successfully. Listening on port 11111.
Finished initializing the daemon package.
Saving notifications settings...`
- I realize this isn't much to go on. Is there any other information I can provide?

### Repro steps

1. Enable SonarLint extension
2. Compile a medium to large project (~ 1 million LoC), repeat if necessary
3. Watch memory consumption increase

### Expected behavior

- no crash, memory freed

### Actual behavior

- memory consumption increases until crash

### Related information

* SonarLint for VS Version: 4.32.0.27625
* Visual Studio Version: Enterprise v16.8.6
* SonarQube server version: 7.9.5 (build 38598)
* Languages in your solution: C#, Javascript

**Comment by duncanp-sonar** [comment]
@bamatroot Firstly, thanks for reporting the issue, for the effort you've put into investigating the impact and for the offer of more info. It's much appreciated.

The issue is intriguing, because I can't think of any specific SonarLint actions that are triggered on build. However, there must be some explanation for the observed behaviour, and we'd definitely like to understand what it is.

A few general questions:

1) "SonarLint disabled": just to clarify, do you mean marking the extension as disabled in "Tools", "Extensions", then restarting VS?
2)  are you using [Connected Mode](https://github.com/SonarSource/sonarlint-visualstudio/wiki/Connected-Mode)? 

Some background, and some more specific questions:

#### JavaScript
Currently our JavaScript analysis per-file, when the file is saved. The analysis work is done out of process, and analysis is not performed on build. VS will automatically save all "dirty" files when a build triggered in the IDE which could trigger multiple JavaScript analyses, but that doesn't seem to match the behaviour you are describing.
Also, our JavaScript analysis can be disabled from "Tools", "Options", "SonarLint", so it should be easy to determine if the JavaScript analysis is a factor.

3) Does the issue still repo with JavaScript analysis disabled?

#### C# analysis
Our C# rules are implemented as Roslyn-based analyzers, and VS is responsible for deciding when they run. By default, Roslyn analyzers that are installed in a VSIX are not run on build, they are only run "live" for the currently open documents i.e. when editing in the IDE. However, if you are using one of the newer VS 2019 updates it is possible to configure VS to run VSIX analyzer rules in a build too by setting RunAnalyzersDuringBuild=true in the project file (see the [MSDocs](https://docs.microsoft.com/en-us/visualstudio/code-quality/disable-code-analysis?view=vs-2019) for more detail).

4a) what [scope](https://docs.microsoft.com/en-us/visualstudio/code-quality/configure-live-code-analysis-scope-managed-code?view=vs-2019) do you have set for live analysis, and (4b) does it make a difference if is set to "current document"?
5) do you have _RunAnalyzersDuringBuild_ set in your projects?
6) do your projects reference the SonarAnalyzer.CSharp NuGet package from your projects? (by default VS will run NuGet analyzers during build)

**Comment by bamatroot** [comment]
@duncanp-sonar Thank you for the quick response.

> 1. "SonarLint disabled": just to clarify, do you mean marking the extension as disabled in "Tools", "Extensions", then restarting VS?

Correct, extension disabled and VS restarted
> 2. are you using Connected Mode?

Yes, I'm using connected mode.
Just tried disconnected (.sonarlint folder renamed).
- memory consumption rises to about 2'500MB, but falls again when starting the next rebuild -> no crashes so far [SonarLint extension still enabled] 

> 3. Does the issue still repo with JavaScript analysis disabled?

Yes, still repos w/ JavaScript analysis disabled
> 4a) what scope do you have set for live analysis, and (4b) does it make a difference if is set to "current document"?

scope is already "current document"

> 5) do you have RunAnalyzersDuringBuild set in your projects?

No, none of the about 50 projects have RunAnalyzersDuringBuild set (to true)
(Each of these projects has it's own output path instead of a centralized one - I'm mentioning this b/c it generates a lot of files - although SonarQube: "Configuration" > "C#" > "File suffixes" is set to *.cs)
> 6) do your projects reference the SonarAnalyzer.CSharp NuGet package from your projects? (by default VS will run NuGet analyzers during build)

No, SonarAnalyzer.CSharp NuGet is not referenced from any project

- During the build of the solution, no new output is written to the SonarLint Output - not sure if this verifies that no SonarLint actions are triggered during build.

**Comment by duncanp-sonar** [comment]
@bamatroot thanks for the info.

> Yes, I'm using connected mode.
Just tried disconnected (.sonarlint folder renamed).

> memory consumption rises to about 2'500MB, but falls again when starting the next rebuild -> no crashes so far [SonarLint extension still enabled]

Ok, so might be related to Connected Mode. 

I haven't been able to reproduce your results locally yet. I'm using the [MSBuild](https://github.com/dotnet/msbuild) code as a sample project. It's not the same size as your solution (a few hundred thousand lines of code) but it should be large enough to see a memory leak in the SonarLint code if there is one. It doesn't have any JavaScript, but that doesn't seem to be the problem anyway. So far the memory is stable at around 1000MB, even after repeated builds. I'll try a different project to see if I get different results.

FYI dump of the managed heap after repeated builds:
![image](https://user-images.githubusercontent.com/10757559/110837077-b8ae4280-8298-11eb-891b-73ca8412d351.png)
There are a hundreds of SonarLint-related classes, but none of them are large in terms of number or memory usage (each C# rule is implemented in a separate class, and there are hundreds of rules).

If you have time, it would be useful to see what the managed heap looks like for your processes once the memory usage has increased. FYI the above screenshot is from the diagnostic tools in VS which I got as follows;
* start the target VS, open the MSBuild solution and build it repeatedly.
* start another VS instance and attach it to the first instance (_Debug, Attach to process, devenv.exe_)
* _Debug, WIndows, Managed memory_
* Click on _View Heap_:
![image](https://user-images.githubusercontent.com/10757559/110838724-b77e1500-829a-11eb-907d-686500d386e3.png)
... then filter by type name "Sonar".

It's also possible to take two snapshots and diff them.

**Comment by bamatroot** [comment]
@duncanp-sonar thanks for your efforts

Overview:
![memory-usage-snapshot-overview](https://user-images.githubusercontent.com/80115332/110949022-675a8d80-8342-11eb-9e8e-d2e4a84cfc05.JPG)
- increases after the first rebuild are small
- Taking snapshots helps w/ freeing memory in time to prevent/delay crashes (during rebuild memory is at max ~ 5'000MB according to TaskManager, and drops to about 4'000MB afterwards. W/o taking snapshots and consecutive rebuilds memory consumption rises way above 5'000 MB and VS crashes)


Snapshot  2 (after first rebuild), filtered by type name "Sonar":
![s2-filtered-sonar](https://user-images.githubusercontent.com/80115332/110949024-67f32400-8342-11eb-936a-453fd1c22619.JPG)
- Similar to your results, sonar objects are small 
- but the PE*Symbol objects go through the roof (screenshot 4). As they're part of Microsoft.CodeAnalysis I suppose there's some analysis going on... I'll double check `RunAnalyzersDuringBuild`, ...


Diff snapshot 1 and 2 (before and after first rebuild), filtered by type name "Sonar":
![diff-s1-s2-filtered-sonar](https://user-images.githubusercontent.com/80115332/110949018-66c1f700-8342-11eb-8e61-697915df7110.JPG)


Diff snapshot 1 and 2 (before and after first rebuild), not filtered:
![diff-s1-s2](https://user-images.githubusercontent.com/80115332/110949027-67f32400-8342-11eb-8aa7-4fa192d7fed8.JPG)

Have a nice weekend.

**Comment by duncanp-sonar** [comment]
@bamatroot that's fantastic, thanks very much for investigating!

### Connected Mode
The `SonarQube.Client.Models.IssueLocation/IssueTextRange` objects are related to Connected Mode. They represent issues that have been fetched from SonarQube so we can match them against locally raised issues (so if an issue has been suppressed on the server we'll try to hide it in the IDE). However, that only accounts for a fraction of the memory usage so it looks like Connected Mode might be a red herring.

### Analysis during build

> the PE*Symbol objects go through the roof (screenshot 4). As they're part of Microsoft.CodeAnalysis I suppose there's some analysis going on...

The build will also create a lot of `Microsoft.CodeAnalysis` objects even if analysis isn't happening. However, you're not getting the huge memory increase if SonarLint is disabled, which suggests that the analyzers are being executed.

#### Build issues in the Error List
How many issues are reported in the Error List if you change the Error List filter to "Build Only"?
![image](https://user-images.githubusercontent.com/10757559/110977796-7c431b00-835a-11eb-9efd-dfbf2ffa2b21.png)

The Error List had (and as far as I know still does) a counter-intuitive behaviour that the default "Build + Intellisense" feature doesn't show the aggregate of build and "live" issues; to see all of the "build" issues you have to change the filter to "Build Only".

#### `RunAnalyzersDuringBuild`
A couple of points to note:
* check for properties called `RunAnalyzers` and `RunAnalyzersDuringBuild` (the first takes precedence over the second)
* the properties could be set in either the "main" `.csproj` file or the `.csproj.user` file. It will be in the `.csproj.user` file if you check the "Run on build" / "Run on live analysis" properties in IDE (see [this MS doc page](https://docs.microsoft.com/en-us/visualstudio/code-quality/disable-code-analysis?view=vs-2019) for more info).


My thinking at the moment is:
* Connected Mode isn't part of the problem
* analysis is being run during build
* a large number of issues is being raised -> large memory consumption.

The fact that memory isn't being freed could be for a number of reasons:
* there are no memory leaks - it just takes that much memory to hold all of the issues that have been raised
* there's a memory leak in the Sonar analyzers (possible - our data-flow rules are memory-intensive and could consume a lot of memory when running, but it should be freed afterwards)
* there's a memory leak in VS/Roslyn.
Or it could be something else entirely :-).

Have a good weekend.

**Comment by bamatroot** [comment]
@duncanp-sonar sorry for the late response - busy week

### `RunAnalyzersDuringBuild`
- text searched all files in the source directory again, no `RunAnalyzersDuringBuild` nor `RunAnalyzers`
- started checking project files in the IDE and by dumb luck one of the first ones I opened had "Run on build" checked
  - unchecking led to the following changes in the corresponding *.csproj: `<RunCodeAnalysis>true</RunCodeAnalysis>` -> `<RunCodeAnalysis>false</RunCodeAnalysis>`
  - sadly, it was the only one
  - (switching to the 'Code Analysis' tab in the project properties takes about 2-4s. But that's the case whether the SonarLint extension is enabled or not)

### Issue count
- about 4K issues are reported in the Error List when filtered to "Build Only"
  - none of them are Sonar issues (only CS...)
  - Sonar issues only get added to the Error List when I switch back to "Build + IntelliSense" and start opening files
  - (total number of Sonar issues for the Sonar project - according to SonarQube is about 44K unresolved + 9K removed, false-positive, ...)

### Misc
- Sometimes when starting VS the yellow performance issue indicator bar appears. "Visual Studio stopped responding for 8 seconds. Disabling the extension SonarLint [...] might help."
- Project properties > Active Rules: there are lots of duplicates of our Sonar way rule set in the combo box - got that number down by using `Directory.Build.props` and removing `<CodeAnalysisRuleSet` from the project files - makes no difference
- checked whether starting VS w/ and w/o admin privileges made any difference - nope

Seems like a tricky issue.

Enjoy your weekend.

**Comment by duncanp-sonar** [comment]
@bamatroot no problem - ditto!

#### Misc

> Sometimes when starting VS the yellow performance issue indicator bar appears. "Visual Studio stopped responding for 8 seconds. Disabling the extension SonarLint [...] might help."

This is a separate known issue (#1318). It only happens on start-up, when there's a race condition between SLVS fetching suppressions from the SonarQube server and VS calling one of our analysis rules to do a live analysis.

#### Issue count
> none of them are Sonar issues (only CS...)

That sounds like what we'd expect. One definitive way of checking which analyzers are executed on build is to check the build output window in VS. Build with build logging set to "Normal":
![image](https://user-images.githubusercontent.com/10757559/112168112-ec6a6000-8be8-11eb-8838-60c68f17a319.png)

... then search for `/analyzer:sonar` in the Build output window pane:
![image](https://user-images.githubusercontent.com/10757559/112167875-b62ce080-8be8-11eb-9547-d7b093154184.png)


One miscellaneous question: your original post contains the log output  `Listening on port 11111`. This isn't a port the released version of SonarLint would ever use. Did you just change the port number in your post so as not to show the real number in a public post?

Thanks.

**Comment by bamatroot** [comment]
@duncanp-sonar
- Output doesn't contain /analyzer:sonar when verbosity is set to "Normal"
- Port 11111 - yes, I changed it.

There are a few things I'll try to do/test next week
- compare build outputs w/ and w/o SonarLint extension enabled - quick comparison showed there are a few hundred lines more w/ SonarLint enabled ...
- update to VS 16.9.2, although there's nothing in the release notes that would get my hopes up...

**Comment by duncanp-sonar** [comment]
@bamatroot 

> Output doesn't contain /analyzer:sonar when verbosity is set to "Normal"
Ok, that confirms the analyzers are definitely not running on build.

> Port 11111
Thanks for confirming.

Have a good weekend.

**Comment by bamatroot** [comment]
Hi @duncanp-sonar 

I updated to VS 16.9.3 and good news: SonarLint extension enabled + Connected Mode -> even after multiple builds memory consumption is about 2GB.

Not sure if the update is the (sole) reason for the issues's disappearence.
- For an unrelated test, I had to delete my local source folder - including the .sonarlint folder. If I remember correctly, there were more files in the folder than there are now. I think there was at a least a \*SonarQube\* file or folder that wasn't created this time. Maybe there were some issues with an older version.

As a side note: The one project with `<RunCodeAnalysis>true</RunCodeAnalysis>` I mentioned in a previous post, had the ["Microsoft Minimum Recommended Rules"](https://docs.microsoft.com/en-us/visualstudio/code-quality/how-to-configure-code-analysis-for-a-managed-code-project?view=vs-2019) assigned, not the SonarLint rule set.

Thank you very much for your assistance and sorry for the "bug".

Have a good weekend.
